### PR TITLE
[feature] Implement custom parameters for plugins

### PR DESCRIPTION
**Pull request by Yanshuming1**
Implement custom parameters for plug-ins

### CODE REVIEW

File: manager/src/main/java/org/apache/hertzbeat/manager/controller/PluginController.java [hunk 1]
```diff
@@ -85,4 +87,19 @@ public ResponseEntity<Message<Void>> updatePluginStatus(@RequestBody PluginMetad
...
     }
+
+    @GetMapping("/getParamDefine")
+    @Operation(summary = "get param define", description = "get param define by jar path")
+    public ResponseEntity<Message<PluginParametersVO>> getParamDefine(@RequestParam Long pluginMetadataId) {
+        PluginParametersVO plugins = pluginService.getParamDefine(pluginMetadataId);
+        return ResponseEntity.ok(Message.success(plugins));
+    }
+
+    @PostMapping("/saveParams")
```
Comment (LiuTianyou) [inline manager/src/main/java/org/apache/hertzbeat/manager/controller/PluginController.java:98]: recommend to use all lowercase letters in the URL and use http method to indicate the action like `/params/define`

File: manager/src/main/java/org/apache/hertzbeat/manager/service/PluginService.java [hunk 2]
```diff
@@ -58,7 +63,7 @@ public interface PluginService {
      * @param execute run plugin logic
      * @param <T> plugin type
      */
-    <T> void pluginExecute(Class<T> clazz, Consumer<T> execute);
+    <T> void pluginExecute(Class<T> clazz, Consumer<T> execute, BiConsumer<T, List<Configmap>> biConsumer);
```
Comment (LiuTianyou) [inline manager/src/main/java/org/apache/hertzbeat/manager/service/PluginService.java:66]: It is recommended to wrap parameters in objects to improve the scalability of the plugin interface and facilitate the subsequent transfer of other data to the plugin for use.

File: manager/src/main/java/org/apache/hertzbeat/manager/service/impl/PluginServiceImpl.java [hunk 3]
```diff
@@ -335,18 +428,30 @@ private List<URL> loadLibInPlugin(String pluginJarPath) {
                         out.flush();
                     }
                 }
+                if ((entry.getName().contains("define")) && (entry.getName().endsWith(".yml") || entry.getName().endsWith(".yaml"))) {
+                    try (InputStream ymlInputStream = jarFile.getInputStream(entry)) {
+                        List<ParamDefine> params = yaml.loadAs(ymlInputStream, List.class);
+                        PARAMS_Define_MAP.putIfAbsent(pluginMetadataId, params);
+                    }
+                }
```
Comment (LiuTianyou) [inline manager/src/main/java/org/apache/hertzbeat/manager/service/impl/PluginServiceImpl.java:436]: It is necessary to add the plugin parameter configuration file content check function to upload Plugin. In my previous test, if define.yml has an error, if the user installs a plugin containing an incorrect configuration file, the manager will not be able to restart.

**Comment by LiuTianyou** [comment]
@Yanshuming1  If click the Edit button, an error message will pop up and the page will become unresponsive.
image

image

File: manager/src/main/java/org/apache/hertzbeat/manager/service/impl/PluginServiceImpl.java [hunk 4]
```diff
@@ -335,18 +428,30 @@ private List<URL> loadLibInPlugin(String pluginJarPath) {
...
             for (T t : load) {
                 if (pluginIsEnable(t.getClass())) {
-                    execute.accept(t);
+                    Long pluginId = ITEM_TO_PLUGINMETADATAID_MAP.get(t.getClass().getName());
+                    List<Configmap> configmapList = PARAMS_MAP.get(pluginId);
+                    if (CollectionUtils.isEmpty(configmapList)) {
+                        execute.accept(t);
+                    } else {
+                        biConsumer.accept(t, configmapList);
+                    }
```
[the remark this answers is not in the document]
Comment (Yanshuming1) [reply manager/src/main/java/org/apache/hertzbeat/manager/service/impl/PluginServiceImpl.java:454]: That is what is being done

**Review by LiuTianyou — APPROVED** [review]
lgtm

**Pull request merged**